README.md
rcusum - An Implementation of Cusum Control Charts as an R package
============
Author: Alexander Meyer

This R Package implements the charts as described by _Rogers et al. (2004, doi:10.1016/j.jtcvs.2004.03.011)_. Cusum control charts are devices for monitoring arbitrary processes for a prespecified event. This package and manual uses the term failure instead of event, although any other type of event such as successes could be monitored as well. The processes of our interest are typically surgical procedures.

To create a basic control chart 4 parameters have to be specified:

* $p0$: risk of failure when the process is in control (acceptable failure rate for the procedure)
* $p1$, where $p1>p0$: failure rate considered unacceptable
* $\alpha$: Type I error (the probability of concluding that the failure rate has increased, when in fact it has not)
* $\beta$: Type II error (the probability of concluding that the failure rate has not increased, when in fact it has)
 
Anatomy of a Cusum Control Chart
--------------------------------
A cusum chart shows the cumulative sum of failures for a given process. There are two boundaries shown on the graph.

If the cusum curve...

* ... lies __within the boundaries__, there is not enough evidence for a significant conclusion, thus there is __still need for monitoring__!
* ... __crossed the upper boundary__, we can conclude that the __failure rate has increased to the unacceptable__ event rate $p1$ (at the specified significance level).
* ... __crossed the lower boundary__, we can conclude that the __failure rate is less than or equal to the acceptable__ event rate $p0$ (at the specified significance level)
 
_"The natural progression of the graph for an individual or institution with acceptable performance is toward the lower boundary for this method of constructing control limits."_  (Rogers et. al, 2004)

rcusum Shiny App
-----------------
You can try some functions of the package interactivly in a Shiny App:
```
require(shiny)
shiny::runGitHub(repo="rcusum",username="meyera", subdir = "inst/rcusum-ShinyApp")
```

NEWS
rcusum 1.0
===========
* added function documentation

* added an ShinyApp to interactivly perform cusum analysis

* new `cusum.obs_minus_exp` function which implements cumulative observed minus expected failure plots for unadjusted and risk adjusted analysis

* new `cusum.sprt` is the risk-adjusted sequential probability ratio test (SPRT) chart with control limits. Basically the risk-adjust counterpart of the cusum log-likelihood chart.

* new `cusum` function implemets cusum chart and cusum log-likelihood chart with control limits for performing formal hypothesis testing


R/functions.R
library(ggplot2)
 
set.seed(16)
common_theme = theme_bw() + theme(axis.line=element_line(colour="black")
                                  ,panel.grid.major=element_blank()
                                  ,legend.position="bottom"
                                  ,legend.title=element_blank()
                                  ,axis.text.x=element_text(angle=30, hjust=1)
                                  ,strip.background=element_rect(fill="grey96"))

df = data.frame(
    is_failure = c(rbinom(50,1,0.10),rbinom(50,1,0.08),rbinom(50,1,0.05),
                   rbinom(50,1,0.10),rbinom(50,1,0.13),rbinom(50,1,0.14),
                   rbinom(50,1,0.14),rbinom(50,1,0.09),rbinom(50,1,0.25)
                   ),
    p0 = c(rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
           rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
           rnorm(50, 0.10, 0.03),rnorm(50, 0.15, 0.03),rnorm(50, 0.20, 0.03)
          ),
    by=rep(factor(c("Surgeon A", "Surgeon B", "Surgeon C")), times=c(150,150,150))
  )

#' Creats a cumulative observed minus expected failure plot
#'
#' Implementation of an unadjusted or risk-adjusted cumulative observed minus expected failure graph as described in Rogers et al. (2004)
#' 
#' "The graph starts at 0, but is incremented by 1 - p0 for a failure and decremented by p0 for a success"
#' 
#' @param failure_indicator a numeric indicator variable consiting of only \code{c(0,1)}, where 0 is no failure and 1 is failure for each procedure
#' @param p0 either a constant which represents the acceptable event rate, or a numeric vector representing the acceptable risk score for each single individual. The later is used when plotting risk audjusted scores. This is then equal to an VLAD (variable life adjusted displays) or CRAM (cumulative risk adjusted mortality) chart 
#' @param by a factor vector consisting of the stratification variable.
#' @param scale_ylim Limits the Y axis scale
#' 
#' @return an object of the class \code{ggplot}
#' 
#' @author Alexander Meyer
#' @family cusum
#' @export
#' 
#' @references Rogers, C. A., Reeves, B. C., Caputo, M., Ganesh, J. S., Bonser, R. S., & Angelini, G. D. (2004). Control chart methods for monitoring cardiac surgical performance and their interpretation Chris. The Journal of Thoracic and Cardiovascular Surgery, 128(6), 811–819. doi:10.1016/j.jtcvs.2004.03.011 
#' 
#' @examples
#' set.seed(16)
#' df = data.frame(
#'  is_failure = c(rbinom(50,1,0.10),rbinom(50,1,0.08),rbinom(50,1,0.05),
#'                  rbinom(50,1,0.10),rbinom(50,1,0.13),rbinom(50,1,0.14),
#'                  rbinom(50,1,0.14),rbinom(50,1,0.09),rbinom(50,1,0.25)
#'  ),
#'  p0 = c(rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.15, 0.03),rnorm(50, 0.20, 0.03)
#'  ),
#'  by=rep(factor(c("Surgeon A", "Surgeon B", "Surgeon C")), times=c(150,150,150))
#' )
#' cp= cusum.obs_minus_exp(rbinom(200,1,0.10), c(rnorm(100, 0.10, 0.03),rnorm(100, 0.10, 0.03)))
#' print(cp)
#' cp= cusum.obs_minus_exp(rbinom(200,1,0.10), 0.10)
#' print(cp)
#' cp= cusum.obs_minus_exp(df$is_failure, 0.10, by=df$by)
#' print(cp)
#' cp= cusum.obs_minus_exp(df$is_failure, df$p0, by=df$by)
#' print(cp)
cusum.obs_minus_exp = function(failure_indicator, p0, by=NULL, scale_ylim = 20) {
  require(ggplot2)
  require(plyr)
  
  stopifnot(is.numeric(failure_indicator))
  stopifnot(failure_indicator %in% c(0,1))
  stopifnot(is.numeric(p0), p0 >= 0 & p0 <= 1 & (length(p0) == length(failure_indicator) | length(p0) == 1))
  stopifnot(is.numeric(scale_ylim), length(scale_ylim) == 1)
  if (!is.null(by)) {
    stopifnot(is.factor(by))
    stopifnot(length(failure_indicator) == length(by))
  }
  
  d = NULL
  if (is.null(by)) {
    failure_line = c()
    for (i in 1:length(failure_indicator)) {
      before = ifelse(i == 1, 0, failure_line[i-1])
      
      p0_ = ifelse(length(p0) == 1, p0, p0[i])
      
      if (failure_indicator[i] == 0) {
        failure_line[i] = before - p0_
      } else if (failure_indicator[i] == 1) {
        failure_line[i] = before + (1-p0_)
      }
    }
    
    d = data.frame(n=1:length(failure_line),
                    failure_line=failure_line)
  } else {
    if (length(p0) == 1) {
      p0_ = rep(p0,times=length(failure_indicator))
    } else {
      p0_ = p0
    }
    
    d = data.frame(
      failure_indicator=failure_indicator,
      by=by,
      p0=p0_
    )
    
    d = ddply(d, .(by), function(sub_d) {  
            failure_line = c()
            for (i in 1:length(sub_d$failure_indicator)) {
              before = ifelse(i == 1, 0, failure_line[i-1])
              
              if (sub_d$failure_indicator[i] == 0) {
                failure_line[i] = before - sub_d$p0[i]
              } else if (sub_d$failure_indicator[i] == 1) {
                failure_line[i] = before + (1-sub_d$p0[i])
              }
            }
            
            data.frame(n=1:length(failure_line),
                       failure_line=failure_line,
                       by=sub_d$by)
          }
        )    
  }
  
  if (is.null(by)) {
    p = ggplot(d, aes(x=n,y=failure_line))
  } else {
    p = ggplot(d, aes(x=n,y=failure_line, linetype=by))
  }
  p = p + geom_hline(yintercept=0, linetype=6, size=1)
  p = p + geom_step()
  
  p = p + ylim(-scale_ylim, scale_ylim)
  p = p + ylab("Cumulative observed minus expected failures")
  p = p + xlab("")
  p <- p + common_theme
  return(p)
}
#cp= cusum.obs_minus_exp(rbinom(200,1,0.10), c(rnorm(100, 0.10, 0.03),rnorm(100, 0.10, 0.03)))
#cp= cusum.obs_minus_exp(rbinom(200,1,0.10), 0.10)
#cp= cusum.obs_minus_exp(df$is_failure, 0.10, by=df$by)
#cp= cusum.obs_minus_exp(df$is_failure, df$p0, by=df$by)
#print(cp + ggthemes::theme_few())


#' Creats a cusum chart and cusum log-likelihood chart
#'
#' Implementation of an unadjusted cusum chart and cusum log-likelihood chart as described in Rogers et al. (2004)
#' 
#' @param failure_indicator a numeric indicator variable consiting of only \code{c(0,1)}, where 0 is no failure and 1 is failure for each procedure
#' @param p0 a constant representing the fixed acceptable event rate when the process is in control
#' @param p1 a constant representing the fixed unacceptable event rate we want to detect
#' @param alpha Type I error (the probability of concluding that the failure rate has increased, when in fact it has not)
#' @param beta Type II error (the probability of concluding that the failure rate has not increased, when in fact it has)
#' @param by a factor vector consisting of the stratification variable.
#' @param loglike_chart a flag controling which kind of chart will be shown
#'
#' @return an object of the class \code{ggplot}
#' 
#' @author Alexander Meyer
#' @family cusum
#' @export
#' 
#' @references Rogers, C. A., Reeves, B. C., Caputo, M., Ganesh, J. S., Bonser, R. S., & Angelini, G. D. (2004). Control chart methods for monitoring cardiac surgical performance and their interpretation Chris. The Journal of Thoracic and Cardiovascular Surgery, 128(6), 811–819. doi:10.1016/j.jtcvs.2004.03.011 
#' 
#' @examples
#' set.seed(16)
#' df = data.frame(
#'  is_failure = c(rbinom(50,1,0.10),rbinom(50,1,0.08),rbinom(50,1,0.05),
#'                  rbinom(50,1,0.10),rbinom(50,1,0.13),rbinom(50,1,0.14),
#'                  rbinom(50,1,0.14),rbinom(50,1,0.09),rbinom(50,1,0.25)
#'  ),
#'  p0 = c(rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.15, 0.03),rnorm(50, 0.20, 0.03)
#'  ),
#'  by=rep(factor(c("Surgeon A", "Surgeon B", "Surgeon C")), times=c(150,150,150))
#' )
#' cusum_plot = cusum(df$is_failure, .10, .20, alpha=0.01,beta=0.01, loglike_chart=TRUE, by=df$by)
#' print(cusum_plot)
#' cusum_plot = cusum(df$is_failure, .10, .20, alpha=0.01,beta=0.01, loglike_chart=FALSE, by=df$by)
#' print(cusum_plot)
cusum = function(failure_indicator, p0, p1, alpha=.01, beta=.01, by=NULL, loglike_chart=FALSE) {
  require(plyr)
  require(ggplot2)
  
  stopifnot(is.numeric(failure_indicator))
  stopifnot(failure_indicator %in% c(0,1))
  stopifnot(is.numeric(p0), p0 >= 0, p0 <= 1)
  stopifnot(is.numeric(p1), p1 >= 0, p1 <= 1)
  stopifnot(is.numeric(alpha), alpha >= 0, alpha <= 1)
  stopifnot(is.numeric(beta), beta >= 0, beta <= 1)
  if (!is.null(by)) {
    stopifnot(is.factor(by))
    stopifnot(length(failure_indicator) == length(by))
  }
    
  OR = (p1*(1-p0))/(p0*(1-p1))
  s  = log((1-p0)/(1-p1))/log(OR)
  h0 = log((1-alpha)/beta)/log(OR)
  h1 = log((1-beta)/alpha)/log(OR)
  n = length(failure_indicator)
  
  if (is.null(by)) {
    d = data.frame(
      n=1:n,
      failure=failure_indicator,
      cusum=cumsum(failure_indicator),
      l0=((1:n)*s)-h0,
      l1=((1:n)*s)+h1,
      T_horizontal = cumsum(failure_indicator - s)
    )  
  } else {
    d = data.frame(
      failure=failure_indicator,
      by=by
    )
    
    d = ddply(d, .(by), function(sub_d) {      
      data.frame(
        n=1:nrow(sub_d),
        failure=sub_d$failure,
        cusum=cumsum(sub_d$failure),
        l0=((1:nrow(sub_d))*s)-h0,
        l1=((1:nrow(sub_d))*s)+h1,
        T_horizontal = cumsum(sub_d$failure - s),
        by=sub_d$by
      )
    })
  }
  
  p = NULL
  if (loglike_chart) {
    d_ = cbind(d, h0=rep(h0,times=nrow(d)), h1=rep(h1, times=nrow(d)))

    if (is.null(by)) {
      p = ggplot(data=d_, aes(y=T_horizontal,x=n))
    } else {
      p = ggplot(data=d_, aes(y=T_horizontal,x=n, linetype=by))
    }
    
    p = p + geom_step() 
    p = p + geom_hline(yintercept=-h0, linetype=2)
    p = p + geom_hline(yintercept=h1, linetype=2)
    p = p + geom_hline(yintercept=0, color="lightgrey")
    
    p = p + ylab("Cumulative log-likelihood ratio")
    
    p <- p + annotate("text", label=paste("Accept H1"), x=-Inf, y=h1*2, hjust=-0.5)#, x=n/5, y=h1*2)
    p <- p + annotate("text", label=paste("Accept H0"), x=-Inf, y=-h0*2, hjust=-0.5)#, x=n/5, y=-h0*2)
  } else {
    
    if (is.null(by)) {
      p = ggplot(data=d ,aes(y=cusum,x=n))
    } else {
      p = ggplot(data=d ,aes(y=cusum,x=n, linetype=by))
    }
    
    p = p + geom_step() 
    p = p + geom_line(mapping=aes(x=n,y=l0),linetype=2)
    p = p + geom_line(mapping=aes(x=n,y=l1),linetype=2)
    
    p <- p + annotate("text", label=paste("Accept H1"), x=-Inf, y=Inf, vjust=1.3, hjust=-0.3)#x=n/3, y=(n/2)*p1*1.0)
    p <- p + annotate("text", label=paste("Accept H0"), x=Inf, y=-Inf, vjust=-1.2, hjust=1.3)#x=n/1.5, y=(n/2)*(p0*1.0))
  }
  
  p <- p + common_theme
  p = p + xlab("")
  
#   if (!is.null(by)) {
#     p <- p + facet_wrap(~by, as.table=TRUE)
#   }
  
  return(p)
}
#cusum_plot = cusum(df$is_failure, .10, .20, alpha=0.01,beta=0.01, loglike_chart=TRUE, by=df$by)
#print(cusum_plot)


#' Creats a risk-adjusted sequential probability ratio test (SPRT) chart
#'
#' Implementation of a risk-adjusted sequential probability ratio test (SPRT) chart with control limits as described in Rogers et al. (2004)
#' 
#' For the unadjusted chart, increase in risk is defined in terms of the unacceptable failure rate. However, when risk for each patient varies, it does not make sense to have a common unacceptable rate applied across all operations.
#' This variable unacceptable rate is achieved by defining the increase in terms of a relative risk (ie. odds ratio), rather than a specific rate.
#' 
#' @param failure_indicator a numeric indicator variable consiting of only \code{c(0,1)}, where 0 is no failure and 1 is failure for each procedure
#' @param p0 a numeric vector representing the acceptable risk score/acceptable failure rate for each single individual. I.e. STS Score values, or emperically modeled risks
#' @param OR the increase in relative risk to the modeled acceptable risk, where An odds ratio of 2, for example, would equate approximately to a doubling of patientspecific risk of failure, an odds ratio of 1.5 to a 50 percent increase in failure risk, and so on.
#' @param alpha Type I error (the probability of concluding that the failure rate has increased, when in fact it has not)
#' @param beta Type II error (the probability of concluding that the failure rate has not increased, when in fact it has)
#' @param by a factor vector consisting of the stratification variable.
#'
#' @return an object of the class \code{ggplot}
#' 
#' @author Alexander Meyer
#' @family cusum
#' @export
#' 
#' @references Rogers, C. A., Reeves, B. C., Caputo, M., Ganesh, J. S., Bonser, R. S., & Angelini, G. D. (2004). Control chart methods for monitoring cardiac surgical performance and their interpretation Chris. The Journal of Thoracic and Cardiovascular Surgery, 128(6), 811–819. doi:10.1016/j.jtcvs.2004.03.011 
#' 
#' @examples
#' set.seed(16)
#' df = data.frame(
#'  is_failure = c(rbinom(50,1,0.10),rbinom(50,1,0.08),rbinom(50,1,0.05),
#'                  rbinom(50,1,0.10),rbinom(50,1,0.13),rbinom(50,1,0.14),
#'                  rbinom(50,1,0.14),rbinom(50,1,0.09),rbinom(50,1,0.25)
#'  ),
#'  p0 = c(rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),rnorm(50, 0.10, 0.03),
#'          rnorm(50, 0.10, 0.03),rnorm(50, 0.15, 0.03),rnorm(50, 0.20, 0.03)
#'  ),
#'  by=rep(factor(c("Surgeon A", "Surgeon B", "Surgeon C")), times=c(150,150,150))
#' )
#' 
#' sprt1= cusum.sprt(rbinom(200,1,0.10), rnorm(200, 0.10, 0.03), 1.5)
#' print(sprt1)
#' sprt2= cusum.sprt(df$is_failure, df$p0, 1.5, by=df$by)
#' print(sprt2)
cusum.sprt = function(failure_indicator, p0, OR, alpha=.01, beta=.01, by=NULL) {
  require(plyr)
  require(ggplot2)
  
  stopifnot(is.numeric(failure_indicator))
  stopifnot(failure_indicator %in% c(0,1))
  stopifnot(is.numeric(p0), p0 >= 0 & p0 <= 1 & length(p0) == length(failure_indicator))
  stopifnot(is.numeric(OR), OR >= 1)
  stopifnot(is.numeric(alpha), alpha >= 0, alpha <= 1)
  stopifnot(is.numeric(beta), beta >= 0, beta <= 1)
  if (!is.null(by)) {
    stopifnot(is.factor(by))
    stopifnot(length(failure_indicator) == length(by))
  }
  
  h0 = log((1-alpha)/beta)/log(OR)
  h1 = log((1-beta)/alpha)/log(OR)
  n = length(failure_indicator)
  
  if (is.null(by)) {
    d = data.frame(
      n=1:n,
      failure=failure_indicator,
      T_horizontal = cumsum(failure_indicator - (log((1-p0)+(OR*p0))/log(OR)))
    )  
  } else {
    d = data.frame(
      failure=failure_indicator,
      p0=p0,
      by=by
    )
    
    d = ddply(d, .(by), function(sub_d) {      
      data.frame(
        n=1:nrow(sub_d),
        failure=sub_d$failure,
        T_horizontal = cumsum(sub_d$failure - (log((1-sub_d$p0)+(OR*sub_d$p0))/log(OR))),
        by=sub_d$by
      )
    })
  }
  
  d_ = cbind(d, h0=rep(h0,times=nrow(d)), h1=rep(h1, times=nrow(d)))
  
  if (is.null(by)) {
    p = ggplot(data=d_, aes(y=T_horizontal,x=n))
  } else {
    p = ggplot(data=d_, aes(y=T_horizontal,x=n, linetype=by))
  }
  
  p = p + geom_step() 
  p = p + geom_hline(yintercept=-h0, linetype=2)
  p = p + geom_hline(yintercept=h1, linetype=2)
  p = p + geom_hline(yintercept=0, color="lightgrey")
  
  p = p + ylab("Cumulative log-likelihood ratio")
  
  p <- p + annotate("text", label=paste("Accept H1"), x=-Inf, y=h1*2, hjust=-0.5)#, x=n/5, y=h1*2)
  p <- p + annotate("text", label=paste("Accept H0"), x=-Inf, y=-h0*2, hjust=-0.5)#, x=n/5, y=-h0*2)
  
  p <- p + common_theme
  p = p + xlab("")
  
  #   if (!is.null(by)) {
  #     p <- p + facet_wrap(~by, as.table=TRUE)
  #   }
  
  return(p)
}
#cp= cusum.obs_minus_exp(rbinom(200,1,0.10), c(rnorm(100, 0.10, 0.03),rnorm(100, 0.10, 0.03)))
#sprt1= cusum.sprt(rbinom(200,1,0.10), rnorm(200, 0.10, 0.03), 1.5)
#sprt2= cusum.sprt(df$is_failure, df$p0, 1.5, by=df$by)
#print(sprt2)

R/RCusum-package.r
#' rcusum - An Implementation of Cusum Control Charts as an R package
#'
#' This R Package implements the charts as described by _Rogers et al. (2004, doi:10.1016/j.jtcvs.2004.03.011)_. Cusum control charts are devices for monitoring arbitrary processes for a prespecified event. This package and manual uses the term failure instead of event, although any other type of event such as successes could be monitored as well. The processes of our interest are typically surgical procedures.
#'
#' @references Rogers, C. A., Reeves, B. C., Caputo, M., Ganesh, J. S., Bonser, R. S., & Angelini, G. D. (2004). Control chart methods for monitoring cardiac surgical performance and their interpretation Chris. The Journal of Thoracic and Cardiovascular Surgery, 128(6), 811–819. doi:10.1016/j.jtcvs.2004.03.011 
#' 
#' @author Alexander Meyer
#' @name rcusum
#' @import ggplot2 plyr
#' @docType package
NULL


inst/rcusum-ShinyApp/server.R
library(shiny)
library(rcusum)
require(gridExtra)

# Define server logic required to generate and plot a random distribution
shinyServer(function(input, output) {
  
  datasetInput <- reactive({
    
    # input$file1 will be NULL initially. After the user selects and uploads a 
    # file, it will be a data frame with 'name', 'size', 'type', and 'datapath' 
    # columns. The 'datapath' column will contain the local filenames where the 
    # data can be found.
    inFile <- input$file1
    
    if (is.null(inFile))
      return(NULL)
    
    read.csv(inFile$datapath, header=input$header, sep=input$sep, quote=input$quote)
  })
  
  output$contents <- renderTable({
    datasetInput()
  })
  
  output$failure_indicator_chooser <- renderUI({
    ind <- names(datasetInput())
    if (is.null(ind) | length(ind) == 0) {
      return(tags$div(style="color:red", "No data uploaded!", tags$br(),"Please upload data."))
    } else {
      return(selectInput("failure_indicator", "Choose Failure Indicator Variable", ind))
    }
  })
  
  output$p0_variable_chooser <- renderUI({
    ind <- names(datasetInput())
    if (is.null(ind) | length(ind) == 0) {
      return(tags$div(style="color:red", "No data uploaded!", tags$br(),"Please upload data."))
    } else {
      return(selectInput("p0_variable", "Choose p0 variable", ind))
    }
  })
  
  output$by_chooser <- renderUI({
    ind <- names(datasetInput())
    if (is.null(ind) | length(ind) == 0) {
      ind = c()
    }
    
    return(selectInput("by_indicator", "Choose a stratification Variable", c("No stratification" ,ind)))
  })
  
  output$unadjusted <- renderPlot({
    df = datasetInput()
    failures = df[[input$failure_indicator]]
    by = NULL
    if (!is.null(input$by_indicator) & input$by_indicator != "No stratification") {
      by = as.factor(df[[input$by_indicator]])
    }
    
    alpha = as.numeric(input$alpha)
    beta = as.numeric(input$beta)
    p0 = as.numeric(input$p0)
    p1 = as.numeric(input$p1)
    
    p1 = cusum(failures, p0=p0, p1=p1, alpha=alpha, beta=beta, by=by, loglike_chart=TRUE)
    p2 = cusum.obs_minus_exp(failures,p0=p0, by=by)

    gridExtra::grid.arrange(p1, p2, nrow=1, as.table=TRUE)
  })
  
  output$risk_adjusted <- renderPlot({
    df = datasetInput()
    failures = df[[input$failure_indicator]]
    p0 = df[[input$p0_variable]]
    by = NULL
    if (!is.null(input$by_indicator) & input$by_indicator != "No stratification") {
      by = as.factor(df[[input$by_indicator]])
    }
    
    alpha = as.numeric(input$alpha)
    beta = as.numeric(input$beta)
    OR = as.numeric(input$OR)
    
    p1 = cusum.sprt(failures, p0, OR, alpha, beta, by)
    p2 = cusum.obs_minus_exp(failures, p0, by)
    
    gridExtra::grid.arrange(p1, p2, nrow=1, as.table=TRUE)
  })
})


inst/rcusum-ShinyApp/ui.R
library(shiny)

# Define UI for application that plots random distributions 
shinyUI(pageWithSidebar(
  
  # Application title
  headerPanel("rcusum - An app for conducting interactive cusum analysis"),
  
  # Sidebar with a slider input for number of observations
  sidebarPanel(
    tags$h4("Data File Upload"),
    fileInput('file1', 'Choose CSV File',
              accept=c('text/csv', 'text/comma-separated-values,text/plain')),
    checkboxInput('header', 'Header', TRUE),
    radioButtons('sep', 'Separator',
                 c(Comma=',',
                   Semicolon=';',
                   Tab='\t'),
                 'Comma'),
    radioButtons('quote', 'Quote',
                 c(None='',
                   'Double Quote'='"',
                   'Single Quote'="'"),
                 'Double Quote'),
    tags$hr(),
    tags$h4("General Cusum Settings"),
    uiOutput("failure_indicator_chooser"),
    uiOutput("by_chooser"),
    tags$br(),
    sliderInput("alpha", 
                "Alpha-Level: Type I error rate", 
                min = 0.001,
                max = 0.2, 
                value = 0.01,
                ticks=TRUE,
                step=0.002
                ),
    sliderInput("beta", 
                "Beta-Level: Type II error rate", 
                min = 0.001,
                max = 0.2, 
                value = 0.01,
                ticks=TRUE,
                step=0.002
    ),
    br(),
    HTML("Please report bugs, errors and comments/suggestions to <a href='mailto:meyer.alexander+github@gmail.com'>me</a>.")
  ),
  
  # Show a plot of the generated distribution
  mainPanel(
    tabsetPanel(
      tabPanel("Data", tableOutput('contents')),
      tabPanel("Unadjusted Cusum Analysis",
               sliderInput("p0", 
                           "p0: The acceptable failure rate", 
                           min = 0.01,
                           max = 0.30, 
                           value = 0.1,
                           ticks=TRUE,
                           step=0.01
               ),
               sliderInput("p1", 
                           "p1: The unacceptable failure rate", 
                           min = 0.01,
                           max = 0.30, 
                           value = 0.2,
                           ticks=TRUE,
                           step=0.01
               ),
               tags$hr(),
               plotOutput("unadjusted")
               ),
      tabPanel("Risk-Adjusted Cusum Analysis",
               uiOutput("p0_variable_chooser"),
               sliderInput("OR", 
                           "OR: The unacceptable realtive increase of failure compared to estimated risk", 
                           min = 1.00,
                           max = 3.00, 
                           value = 1.5,
                           ticks=TRUE,
                           step=0.01
               ),
               plotOutput("risk_adjusted")
               ),
      tabPanel("Help", tags$a(href="https://github.com/meyera/rcusum", "Go to the package development website for help"))
    )
  )
))